src/libcore/array/iter.rs

```diff
@@ -0,0 +1,266 @@
+    /// - `alive.end <= N`
+    alive: Range<usize>,
+}
+
+impl<T, const N: usize> IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    /// Creates a new iterator over the given `array`.
+    ///
+    /// *Note*: this method might never get stabilized and/or removed in the
+    /// future as there will likely be another, preferred way of obtaining this
+    /// iterator (either via `IntoIterator` for arrays or via another way).
+    #[unstable(feature = "array_value_iter", issue = "0")]
+    pub fn new(array: [T; N]) -> Self {
+        // The transmute here is actually safe. The docs of `MaybeUninit`
+        // promise:
+        //
+        // > `MaybeUninit<T>` is guaranteed to have the same size and alignment
+        // > as `T`.
+        //
+        // The docs even show a transmute from an array of `MaybeUninit<T>` to
+        // an array of `T`.
+        //
+        // With that, this initialization satisfies the invariants.
+
+        // FIXME(LukasKalbertodt): actually use `mem::transmute` here, once it
+        // works with const generics:
+        //     `mem::transmute::<[T; {N}], [MaybeUninit<T>; {N}]>(array)`
+        //
+        // Until then, we do it manually here. We first create a bitwise copy
+        // but cast the pointer so that it is treated as a different type. Then
+        // we forget `array` so that it is not dropped.
+        let data = unsafe {
+            let data = ptr::read(&array as *const [T; N] as *const [MaybeUninit<T>; N]);
+            mem::forget(array);
+            data
+        };
+
+        Self {
+            data,
+            alive: 0..N,
+        }
+    }
+
+    /// Returns an immutable slice of all elements that have not been yielded
+    /// yet.
+    fn as_slice(&self) -> &[T] {
+        // This transmute is safe. As mentioned in `new`, `MaybeUninit` retains
+        // the size and alignment of `T`. Furthermore, we know that all
+        // elements within `alive` are properly initialized.
+        let slice = &self.data[self.alive.clone()];
+        unsafe {
+            mem::transmute::<&[MaybeUninit<T>], &[T]>(slice)
+        }
+    }
+}
+
+
+#[stable(feature = "array_value_iter_impls", since = "1.38.0")]
+impl<T, const N: usize> Iterator for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    type Item = T;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.alive.start == self.alive.end {
+            return None;
+        }
+
+        // Bump start index.
+        //
+        // From the check above we know that `alive.start != alive.end`.
+        // Combine this with the invariant `alive.start <= alive.end`, we know
+        // that `alive.start < alive.end`. Increasing `alive.start` by 1
+        // maintains the invariant regarding `alive`. However, due to this
+        // change, for a short time, the alive zone is not `data[alive]`
+        // anymore, but `data[idx..alive.end]`.
+        let idx = self.alive.start;
+        self.alive.start += 1;
```

Comment:
All of this is just `let idx = self.alive.next()?;`, might be worth simplifying?

Reply:
Good point. However, I think I will keep it like that for now. It's not a whole lot longer, but a lot clearer at first glance, which I think is important when dealing with `unsafe` stuff.